File: backend/app/models/layout.py
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import (
    Boolean,
    DateTime,
    Float,
    ForeignKey,
    Integer,
    JSON,
    String,
    UniqueConstraint,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database.base import Base


if TYPE_CHECKING:
    from app.models.line import Line
    from app.models.layout_machine import LayoutMachine


class Layout(Base):
    __tablename__ = "layouts"

    __table_args__ = (
        UniqueConstraint(
            "line_id",
            "layout_year",
            "layout_month",
            "layout_number",
            name="uq_layout_line_month_number",
        ),
    )

    id: Mapped[int] = mapped_column(
        primary_key=True
    )

    line_id: Mapped[int] = mapped_column(
        ForeignKey("lines.id"),
        nullable=False,
    )

    # ========================================================
    # BASIC LAYOUT INFORMATION
    # ========================================================

    style: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
    )

    job: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
    )

    buyer: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )

    smv: Mapped[float] = mapped_column(
        Float,
        nullable=False,
    )

    # ========================================================
    # MONTHLY LAYOUT NUMBERING
    # ========================================================

    # Example:
    # Line 5 - September 2026:
    #   layout_number = 1
    #   layout_number = 2
    #   layout_number = 3
    #
    # Line 5 - October 2026:
    #   layout_number = 1

    layout_number: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )

    layout_year: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )

    layout_month: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )

    # ========================================================
    # MACHINE INFORMATION
    # ========================================================

    required_machine_count: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
    )

    total_machines: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
    )

    machine_status: Mapped[dict] = mapped_column(
        JSON,
        nullable=False,
    )

    # ========================================================
    # LAYOUT STATUS
    # ========================================================

    status: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
        default="RUNNING",
    )

    # ========================================================
    # TIME INFORMATION
    # ========================================================

    started_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.utcnow,
        nullable=False,
    )

    completed_at: Mapped[datetime | None] = mapped_column(
        DateTime,
        nullable=True,
    )

    duration_minutes: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
    )

    # ========================================================
    # GENERAL INFORMATION
    # ========================================================

    is_active: Mapped[bool | None] = mapped_column(
        Boolean,
        nullable=True,
    )

    notes: Mapped[str | None] = mapped_column(
        String(500),
        nullable=True,
    )

    created_at: Mapped[datetime | None] = mapped_column(
        DateTime,
        nullable=True,
    )

    updated_at: Mapped[datetime | None] = mapped_column(
        DateTime,
        nullable=True,
    )

    # ========================================================
    # RELATIONSHIPS
    # ========================================================

    line: Mapped["Line"] = relationship(
        "Line",
        back_populates="layouts",
    )

    machines: Mapped[list["LayoutMachine"]] = relationship(
        "LayoutMachine",
        back_populates="layout",
        cascade="all, delete-orphan",
        order_by="LayoutMachine.machine_number",
    )

File: web/src/types/layout.ts
export interface Layout {
  id: number;
  line_id: number;
  style?: string | null;
  job?: string | null;
  buyer: string;
  smv: number;
  required_machine_count?: number | null;
  total_machines: number;
  machine_status: Record<string, string>;
  status: string;
  started_at: string;
  completed_at?: string | null;
  duration_minutes?: number | null;
  is_active?: boolean | null;
  notes?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface LayoutCreate {
  line_id: number;
  style?: string | null;
  job?: string | null;
  buyer: string;
  smv: number;
  required_machine_count?: number | null;
  total_machines: number;
  machine_status: Record<string, string>;
  status?: string;
  notes?: string | null;
}

export interface LayoutUpdate {
  style?: string | null;
  job?: string | null;
  buyer?: string | null;
  smv?: number | null;
  required_machine_count?: number | null;
  total_machines?: number | null;
  machine_status?: Record<string, string> | null;
  status?: string | null;
  notes?: string | null;
}

export interface LayoutMachine {
  id: number;
  layout_id: number;
  machine_number: number;
  machine_name?: string | null;
  status: string;
  reason?: string | null;
  is_active: boolean;
  started_at?: string | null;
  completed_at?: string | null;
}

export interface LayoutMachineCreate {
  layout_id: number;
  machine_number: number;
  machine_name?: string | null;
  status?: string;
  reason?: string | null;
}

export interface LayoutMachineUpdate {
  machine_name?: string | null;
  status?: string | null;
  reason?: string | null;
  is_active?: boolean | null;
}